File: src/main/java/database/DatabaseManager.java
package database;

import java.util.ArrayList;
import java.util.List;
import javax.annotation.Nonnull;
import javax.annotation.Nullable;
import javax.servlet.ServletContext;
import sensormodels.DatabaseModel;

public final class DatabaseManager implements DbManager<DatabaseModel>, DatabasePublisher {

  private static DatabaseManager instance;
  private final ArrayList<DbManager> databases = new ArrayList<>();

  public DatabaseManager(ServletContext servletContext) {
    init(servletContext);
  }

  /**
   * Get singleton instance of DatabaseManager.
   *
   * @param servletContext servlet context
   * @return singleton instance
   */
  public static DatabaseManager getInstance(ServletContext servletContext) {
    if (instance == null) {
      instance = new DatabaseManager(servletContext);
    }
    return instance;
  }

  @Override
  public void init(@Nullable ServletContext servletContext) {
    databases.forEach(database -> database.init(servletContext));
  }

  @Override
  public <V extends DatabaseModel> void insertSensorDataList(@Nonnull List<V> sensorDataList) {
    for (DbManager dbManager : databases) {
      try {
        dbManager.insertSensorDataList(sensorDataList);
      } catch (Exception e) {
        e.printStackTrace();
      }
    }
  }

  @Override
  public <V extends DatabaseModel> void insertSensorData(V sensorData) {
    for (DbManager dbManager : databases) {
      try {
        dbManager.insertSensorData(sensorData);
      } catch (Exception e) {
        e.printStackTrace();
      }
    }
  }

  @Override
  public void addDatabase(DbManager dbManager) {
    this.databases.add(dbManager);
  }

  @Override
  public void removeDatabase(DbManager dbManager) {
    this.databases.remove(dbManager);
  }

  @Override
  public boolean hasDatabaseManager(DbManager dbManager) {
    return this.databases.contains(dbManager);
  }
}

File: src/main/java/database/DatabaseManager.kt
package database

import sensormodels.DatabaseModel
import sensormodels.store.models.SuperStoreModel
import java.util.*
import javax.annotation.Nonnull
import javax.servlet.ServletContext

class DatabaseManager(servletContext: ServletContext?) : DbManager<DatabaseModel>,
    DatabasePublisher {
    private val databases = ArrayList<DbManager<DatabaseModel>>()
    override fun init(servletContext: ServletContext?) {
        databases.apply {
            add(MongoDbManager.instance as DbManager<DatabaseModel>)
            add(LuceneManager.getInstance(servletContext)!! as DbManager<DatabaseModel>)
            add(MySqlManager.instance!! as DbManager<DatabaseModel>)
            forEach { database -> database.init(servletContext) }
        }
    }

    override fun insertSensorDataList(@Nonnull sensorDataList: List<DatabaseModel>) {
        databases.forEach {
            try {
                it.insertSensorDataList(sensorDataList)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    override fun insertSensorData(sensorData: DatabaseModel) {
        databases.forEach {
            try {
                it.insertSensorData(sensorData)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }
    }

    override fun addDatabase(dbManager: DbManager<DatabaseModel>) {
        databases.add(dbManager)
    }

    override fun removeDatabase(dbManager: DbManager<DatabaseModel>) {
        databases.remove(dbManager)
    }

    override fun hasDatabaseManager(dbManager: DbManager<DatabaseModel>): Boolean {
        return dbManager in databases
    }

    companion object {
        private var instance: DatabaseManager? = null

        /**
         * Get singleton instance of DatabaseManager.
         *
         * @param servletContext servlet context
         * @return singleton instance
         */
        @JvmStatic
        fun getInstance(servletContext: ServletContext?): DbManager<SuperStoreModel> {
            if (instance == null) {
                instance = DatabaseManager(servletContext)
            }
            return instance as DbManager<SuperStoreModel>
        }
    }

    init {
        init(servletContext)
    }
}